File: 백준/Silver/2531. 회전 초밥/회전 초밥.cc
#include <bits/stdc++.h>

using namespace std;

int n, d, k, c;
int arr[30001];
int visited[30001];
int Max;

int main()
{
	cin >> n >> d >> k >> c;

	// 배열은 컨베이어 벨트이므로 처음과 끝이 이어져있음.
	for (int i = 0; i < n; i++)
	{
		cin >> arr[i];
	}


	// 연속된 k개 초밥이 종류가 가장 다양해야 하며
	// c가 안들어 있는 조합으로 선택해야 최고의 효율을 보임

	// 접시수 n 초밥 가짓 수 d 연속해서 먹는 수 k 쿠폰 번호 c

	for (int i = 0; i < n; i++)
	{
		int eat = 0;
		for (int j = i; j < i + k; j++)
		{
			if (visited[arr[j%n]] == 0)
			{
				eat++;
				visited[arr[j%n]] = 1;
			}
		}

		if (eat >= Max)
		{
			if (visited[c] == 1)
			{
				Max = eat;
			}
			else
			{
				Max = eat + 1;
			}
		}

		for (int j = i; j < i + k; j++)
		{
			visited[arr[j%n]] = 0;
		}
	}

	cout << Max << endl;

	return 0;
}

File: 백준/Silver/2531. 회전 초밥/회전 초밥.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, d, k, c, max;
    static int[] arr, sushi;
    static int[] visited;

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        d = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());

        arr = new int[n];
        sushi = new int[2 * n - 1];
        visited = new int[d + 1];

        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            sushi[i] = arr[i];
        }

        for (int i = n; i < sushi.length ; i++) {
            sushi[i] = arr[i - n];
        }

        int left = 0;
        int right = k-1;
        int count = 0;

        for (int i = 0 ; i < k ; i ++) {
            if (visited[sushi[i]] == 0) {
                count ++;
            }
            visited[sushi[i]] ++;
        }

        while(right < sushi.length) {

            // 확인 먼저 해주고
            if (count >= max) {
                if (visited[c] > 0) {
                    max = count;
                } else {
                    max = count + 1;
                }
            }

            // right, left 옮겨주고
            if (visited[sushi[left]] == 1) {
                count --;
            }
            visited[sushi[left]] --;
            left ++;

            right ++;
            if (right < sushi.length) {
                visited[sushi[right]] ++;
                if (visited[sushi[right]] == 1) {
                    count ++;
                }
            }
        }

        System.out.println(max);
    }
}